pat/consumer_tests/step_4.py

```diff
@@ -0,0 +1,469 @@
+                    f'periodicity defined in the MDIB (but at most {max_self_check_period} seconds) for '
+                    f'"{alert_system.Handle}" within {alert_system.SelfCheckPeriod} '
+                    f'(+{network_delay}s network delay) seconds.',
+                )
+            else:
+                result_collector.ResultCollector.log_failure(
+                    step=step,
+                    message=f'The reference provider produced alert system self check updates for '
+                    f'"{alert_system.Handle}" with a duration of {duration:.2f} seconds, which is not in '
+                    f'accordance to the periodicity defined in the MDIB {alert_system.SelfCheckPeriod} seconds '
+                    f'(and at most {max_self_check_period} seconds).',
+                )
+
+
+def _on_waveform_updates(waveform_updates: dict, waveforms_by_handle: dict):
+    for handle, state in waveforms_by_handle.items():
+        waveform_updates[handle].append(state)
+
+
+def _verify_waveform_tests(  # noqa: PLR0913
+    mdib: ConsumerMdib,
+    step: str,
+    at_least_waveform_descriptors: int,
+    waveform_updates_per_second: int,
+    samples_per_message: int,
+    timeout: float,
+    network_delay: float,
+):
+    waveform_descriptors = mdib.descriptions.NODETYPE.get(pm_qnames.RealTimeSampleArrayMetricDescriptor)
+    if len(waveform_descriptors) < at_least_waveform_descriptors:
+        result_collector.ResultCollector.log_failure(
+            step=step,
+            message=f'The reference provider does not provide at least {at_least_waveform_descriptors} '
+            f'RealTimeSampleArrayMetricDescriptor, but {len(waveform_descriptors)}.',
+        )
+        return
+
+    waveform_updates: dict[str, list[statecontainers.RealTimeSampleArrayMetricStateContainer]] = (
+        collections.defaultdict(list)
+    )
+    observer = functools.partial(_on_waveform_updates, waveform_updates)
+    with observables.bound_context(mdib, waveform_by_handle=observer):
+        time.sleep(timeout + network_delay)
+    if len(waveform_updates) < at_least_waveform_descriptors:
+        result_collector.ResultCollector.log_failure(
+            step=step,
+            message=f'The reference provider did not produce updates for at least {at_least_waveform_descriptors} '
+            f'waveforms, but {len(waveform_updates)}.',
+        )
+
+    # Track how many waveforms meet the criteria
+    # check for equality is not reliable due to network delays
+    waveforms_with_sufficient_updates = sum(
+        1 for updates in waveform_updates.values() if len(updates) >= waveform_updates_per_second * timeout
+    )
+    # check if all updates have at least samples_per_message samples.
+    # due to network delays, we cannot guarantee that all updates have exactly samples_per_message samples
+    waveforms_with_sufficient_samples = sum(
```

Comment:
waveforms_with_sufficient_samples and waveforms_with_sufficient_updates should be verified in one list comprehension